{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Processing MAGs for Model Usage"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import os\n",
    "import requests\n",
    "import re\n",
    "import sys\n",
    "\n",
    "sys.path.append(\"..\")\n",
    "DATA_DIR = \"C:/Users/jakel/Desktop/code/data/\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## For KO-annotated MAGs:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load ko-annotated MAG (LCHF)\n",
    "mag = pd.read_csv(os.path.join(DATA_DIR, \"Thermodesulfo_MAG.csv\"))\n",
    "mag = mag.dropna()\n",
    "ko_list = mag[\"KO\"].to_list()\n",
    "\n",
    "def fetch_and_extract_ec_numbers(ko):\n",
    "    url = f'https://rest.kegg.jp/get/ko:{ko}'\n",
    "    response = requests.get(url)\n",
    "    \n",
    "    if response.status_code == 200:\n",
    "        data = response.text\n",
    "        ec_numbers = re.findall(r'\\b\\d+\\.\\d+\\.\\d+\\.\\d+\\b', data)\n",
    "        return ec_numbers\n",
    "    else:\n",
    "        print(f\"Failed to retrieve data for KO: {ko}\")\n",
    "        return []\n",
    "\n",
    "all_ec_numbers = {}\n",
    "for ko in ko_list:\n",
    "    ec_numbers = fetch_and_extract_ec_numbers(ko)\n",
    "    all_ec_numbers[ko] = ec_numbers\n",
    "\n",
    "data_for_df = []\n",
    "for ko, ec_numbers in all_ec_numbers.items():\n",
    "    for ec in ec_numbers:\n",
    "        data_for_df.append({'KO': ko, 'EC Number': ec})\n",
    "df = pd.DataFrame(data_for_df)\n",
    "\n",
    "cp = df.copy()\n",
    "x = \"LC_bin_0\"\n",
    "y = \"unknown\"\n",
    "z = \"Thermodesulfovibrionales\"\n",
    "s = \"MAG\"\n",
    "cp[[\"taxon_id\",\"media_id\", \"species\", \"source\"]] = x, y, z, s\n",
    "cp = cp.rename(columns={\"EC Number\":\"ec\"})\n",
    "\n",
    "cp.to_csv(os.path.join(DATA_DIR, \"MAGs\", \"test.csv\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Merge our results\n",
    "taxa = pd.read_csv(os.path.join(DATA_DIR, \"taxa2ec-final.csv\"), low_memory=False)\n",
    "cp = pd.read_csv(os.path.join(DATA_DIR, \"MAGs\", \"test.csv\"))\n",
    "test = cp[[\"media_id\", \"species\", \"taxon_id\", \"source\", \"ec\"]]\n",
    "data = [test, taxa]\n",
    "model = pd.concat(data)\n",
    "\n",
    "#model.to_csv(os.path.join(DATA_DIR, \"test_model.csv\"), index=False)\n",
    "model.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## For parsing EC-annotated .gff's':"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processing file: C:/Users/jakel/Desktop/code/data/MAGs\\FENIX21\\DRAM\\all\\KBase_derived_Bin.001.fastaBA_F_extracted_bins.AssemblySet_DRAM.gff\n",
      "Processing file: C:/Users/jakel/Desktop/code/data/MAGs\\FENIX21\\DRAM\\all\\KBase_derived_Bin.002.fastaBA_F_extracted_bins.AssemblySet_DRAM.gff\n",
      "Processing file: C:/Users/jakel/Desktop/code/data/MAGs\\FENIX21\\DRAM\\all\\KBase_derived_Bin.004.fastaBA_F_extracted_bins.AssemblySet_DRAM.gff\n",
      "Processing file: C:/Users/jakel/Desktop/code/data/MAGs\\FENIX21\\DRAM\\all\\KBase_derived_Bin.005.fastaBA_F_extracted_bins.AssemblySet_DRAM.gff\n",
      "Processing file: C:/Users/jakel/Desktop/code/data/MAGs\\FENIX21\\DRAM\\all\\KBase_derived_Bin.006.fastaBA_F_extracted_bins.AssemblySet_DRAM.gff\n",
      "Processing file: C:/Users/jakel/Desktop/code/data/MAGs\\FENIX21\\DRAM\\all\\KBase_derived_Bin.008.fastaBA_F_extracted_bins.AssemblySet_DRAM.gff\n",
      "Processing file: C:/Users/jakel/Desktop/code/data/MAGs\\FENIX21\\DRAM\\all\\KBase_derived_Bin.009.fastaBA_F_extracted_bins.AssemblySet_DRAM.gff\n",
      "Processing file: C:/Users/jakel/Desktop/code/data/MAGs\\FENIX21\\DRAM\\all\\KBase_derived_Bin.011.fastaBA_F_extracted_bins.AssemblySet_DRAM.gff\n",
      "Processing file: C:/Users/jakel/Desktop/code/data/MAGs\\FENIX21\\DRAM\\all\\KBase_derived_Bin.012.fastaBA_F_extracted_bins.AssemblySet_DRAM.gff\n",
      "Processing file: C:/Users/jakel/Desktop/code/data/MAGs\\FENIX21\\DRAM\\all\\KBase_derived_Bin.013.fastaBA_F_extracted_bins.AssemblySet_DRAM.gff\n",
      "Processing file: C:/Users/jakel/Desktop/code/data/MAGs\\FENIX21\\DRAM\\all\\KBase_derived_Bin.014.fastaBA_F_extracted_bins.AssemblySet_DRAM.gff\n",
      "Processing file: C:/Users/jakel/Desktop/code/data/MAGs\\FENIX21\\DRAM\\all\\KBase_derived_Bin.015.fastaBA_F_extracted_bins.AssemblySet_DRAM.gff\n",
      "Processing file: C:/Users/jakel/Desktop/code/data/MAGs\\FENIX21\\DRAM\\all\\KBase_derived_Bin.016.fastaBA_F_extracted_bins.AssemblySet_DRAM.gff\n",
      "Processing file: C:/Users/jakel/Desktop/code/data/MAGs\\FENIX21\\DRAM\\all\\KBase_derived_Bin.017.fastaBA_F_extracted_bins.AssemblySet_DRAM.gff\n",
      "Processing file: C:/Users/jakel/Desktop/code/data/MAGs\\FENIX21\\DRAM\\all\\KBase_derived_Bin.018.fastaBA_F_extracted_bins.AssemblySet_DRAM.gff\n",
      "Processing file: C:/Users/jakel/Desktop/code/data/MAGs\\FENIX21\\DRAM\\all\\KBase_derived_Bin.020.fastaBA_F_extracted_bins.AssemblySet_DRAM.gff\n",
      "Processing file: C:/Users/jakel/Desktop/code/data/MAGs\\FENIX21\\DRAM\\all\\KBase_derived_Bin.021.fastaBA_F_extracted_bins.AssemblySet_DRAM.gff\n",
      "Processing file: C:/Users/jakel/Desktop/code/data/MAGs\\FENIX21\\DRAM\\all\\KBase_derived_Bin.022.fastaBA_F_extracted_bins.AssemblySet_DRAM.gff\n",
      "Processing file: C:/Users/jakel/Desktop/code/data/MAGs\\FENIX21\\DRAM\\all\\KBase_derived_Bin.023.fastaBA_F_extracted_bins.AssemblySet_DRAM.gff\n",
      "Processing file: C:/Users/jakel/Desktop/code/data/MAGs\\FENIX21\\DRAM\\all\\KBase_derived_Bin.024.fastaBA_F_extracted_bins.AssemblySet_DRAM.gff\n",
      "Processing file: C:/Users/jakel/Desktop/code/data/MAGs\\FENIX21\\DRAM\\all\\KBase_derived_Bin.025.fastaBA_F_extracted_bins.AssemblySet_DRAM.gff\n",
      "Processing file: C:/Users/jakel/Desktop/code/data/MAGs\\FENIX21\\DRAM\\all\\KBase_derived_Bin.026.fastaBA_F_extracted_bins.AssemblySet_DRAM.gff\n",
      "Processing file: C:/Users/jakel/Desktop/code/data/MAGs\\FENIX21\\DRAM\\all\\KBase_derived_Bin.028.fastaBA_F_extracted_bins.AssemblySet_DRAM.gff\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>filename</th>\n",
       "      <th>seqname</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>ID</th>\n",
       "      <th>product</th>\n",
       "      <th>ec</th>\n",
       "      <th>ko</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>KBase_derived_Bin.001.fastaBA_F_extracted_bins...</td>\n",
       "      <td>NODE_1052_length_11999_cov_4.021266</td>\n",
       "      <td>KBase</td>\n",
       "      <td>gene</td>\n",
       "      <td>NODE_1052_length_11999_cov_4.021266_4</td>\n",
       "      <td>release factor glutamine methyltransferase %5B...</td>\n",
       "      <td>2.1.1.297</td>\n",
       "      <td>KO:K02493</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>KBase_derived_Bin.001.fastaBA_F_extracted_bins...</td>\n",
       "      <td>NODE_1052_length_11999_cov_4.021266</td>\n",
       "      <td>KBase</td>\n",
       "      <td>gene</td>\n",
       "      <td>NODE_1052_length_11999_cov_4.021266_6</td>\n",
       "      <td>histidyl-tRNA synthetase %5BEC:6.1.1.21%5D</td>\n",
       "      <td>6.1.1.21</td>\n",
       "      <td>KO:K01892</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>KBase_derived_Bin.001.fastaBA_F_extracted_bins...</td>\n",
       "      <td>NODE_1109_length_11407_cov_3.999383</td>\n",
       "      <td>KBase</td>\n",
       "      <td>gene</td>\n",
       "      <td>NODE_1109_length_11407_cov_3.999383_3</td>\n",
       "      <td>type III pantothenate kinase %5BEC:2.7.1.33%5D</td>\n",
       "      <td>2.7.1.33</td>\n",
       "      <td>KO:K03525</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>KBase_derived_Bin.001.fastaBA_F_extracted_bins...</td>\n",
       "      <td>NODE_1114_length_11353_cov_4.046645</td>\n",
       "      <td>KBase</td>\n",
       "      <td>gene</td>\n",
       "      <td>NODE_1114_length_11353_cov_4.046645_8</td>\n",
       "      <td>tRNA-uridine 2-sulfurtransferase %5BEC:2.8.1.1...</td>\n",
       "      <td>2.8.1.13</td>\n",
       "      <td>KO:K00566</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>KBase_derived_Bin.001.fastaBA_F_extracted_bins...</td>\n",
       "      <td>NODE_1141_length_11155_cov_4.291892</td>\n",
       "      <td>KBase</td>\n",
       "      <td>gene</td>\n",
       "      <td>NODE_1141_length_11155_cov_4.291892_2</td>\n",
       "      <td>branched-chain amino acid aminotransferase %5B...</td>\n",
       "      <td>2.6.1.42</td>\n",
       "      <td>KO:K00826</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                             filename  \\\n",
       "3   KBase_derived_Bin.001.fastaBA_F_extracted_bins...   \n",
       "5   KBase_derived_Bin.001.fastaBA_F_extracted_bins...   \n",
       "13  KBase_derived_Bin.001.fastaBA_F_extracted_bins...   \n",
       "29  KBase_derived_Bin.001.fastaBA_F_extracted_bins...   \n",
       "36  KBase_derived_Bin.001.fastaBA_F_extracted_bins...   \n",
       "\n",
       "                                seqname source feature  \\\n",
       "3   NODE_1052_length_11999_cov_4.021266  KBase    gene   \n",
       "5   NODE_1052_length_11999_cov_4.021266  KBase    gene   \n",
       "13  NODE_1109_length_11407_cov_3.999383  KBase    gene   \n",
       "29  NODE_1114_length_11353_cov_4.046645  KBase    gene   \n",
       "36  NODE_1141_length_11155_cov_4.291892  KBase    gene   \n",
       "\n",
       "                                       ID  \\\n",
       "3   NODE_1052_length_11999_cov_4.021266_4   \n",
       "5   NODE_1052_length_11999_cov_4.021266_6   \n",
       "13  NODE_1109_length_11407_cov_3.999383_3   \n",
       "29  NODE_1114_length_11353_cov_4.046645_8   \n",
       "36  NODE_1141_length_11155_cov_4.291892_2   \n",
       "\n",
       "                                              product         ec         ko  \n",
       "3   release factor glutamine methyltransferase %5B...  2.1.1.297  KO:K02493  \n",
       "5          histidyl-tRNA synthetase %5BEC:6.1.1.21%5D   6.1.1.21  KO:K01892  \n",
       "13     type III pantothenate kinase %5BEC:2.7.1.33%5D   2.7.1.33  KO:K03525  \n",
       "29  tRNA-uridine 2-sulfurtransferase %5BEC:2.8.1.1...   2.8.1.13  KO:K00566  \n",
       "36  branched-chain amino acid aminotransferase %5B...   2.6.1.42  KO:K00826  "
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import modules.utils as utils\n",
    "\n",
    "directory = os.path.join(DATA_DIR, 'MAGs', 'FENIX21', 'DRAM', 'all') #can also separate sediments / fluids\n",
    "bins = utils.process_directory(directory) # Parse .gff files for ec numbers\n",
    "\n",
    "bins.to_csv(os.path.join(DATA_DIR, 'MAGs', 'FENIX21', 'FENIX21-bins.csv'), index=False)\n",
    "\n",
    "test_ecs = bins.copy()\n",
    "filter = test_ecs[\"ec\"].str.contains(\"N/A\")\n",
    "test_ecs = test_ecs[~filter]\n",
    "test_ecs.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>media_id</th>\n",
       "      <th>species</th>\n",
       "      <th>taxon_id</th>\n",
       "      <th>source</th>\n",
       "      <th>ec</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1a</td>\n",
       "      <td>Comamonas testosteroni</td>\n",
       "      <td>1886637.0</td>\n",
       "      <td>uniprot</td>\n",
       "      <td>2.6.1.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1a</td>\n",
       "      <td>Comamonas testosteroni</td>\n",
       "      <td>1886637.0</td>\n",
       "      <td>uniprot</td>\n",
       "      <td>4.1.1.12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1a</td>\n",
       "      <td>Comamonas testosteroni</td>\n",
       "      <td>1886637.0</td>\n",
       "      <td>uniprot</td>\n",
       "      <td>1.13.11.74</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1a</td>\n",
       "      <td>Comamonas testosteroni</td>\n",
       "      <td>1886637.0</td>\n",
       "      <td>uniprot</td>\n",
       "      <td>1.13.11.76</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1a</td>\n",
       "      <td>Comamonas testosteroni</td>\n",
       "      <td>1886637.0</td>\n",
       "      <td>uniprot</td>\n",
       "      <td>1.14.13.23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42861</th>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>KBase_derived_Bin.028.fastaBA_F_extracted_bins...</td>\n",
       "      <td>KBase</td>\n",
       "      <td>2.7.1.36</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42867</th>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>KBase_derived_Bin.028.fastaBA_F_extracted_bins...</td>\n",
       "      <td>KBase</td>\n",
       "      <td>6.3.4.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42869</th>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>KBase_derived_Bin.028.fastaBA_F_extracted_bins...</td>\n",
       "      <td>KBase</td>\n",
       "      <td>1.5.98.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42872</th>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>KBase_derived_Bin.028.fastaBA_F_extracted_bins...</td>\n",
       "      <td>KBase</td>\n",
       "      <td>2.7.4.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42876</th>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>KBase_derived_Bin.028.fastaBA_F_extracted_bins...</td>\n",
       "      <td>KBase</td>\n",
       "      <td>2.4.2.48</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1670291 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      media_id                 species  \\\n",
       "0           1a  Comamonas testosteroni   \n",
       "1           1a  Comamonas testosteroni   \n",
       "2           1a  Comamonas testosteroni   \n",
       "3           1a  Comamonas testosteroni   \n",
       "4           1a  Comamonas testosteroni   \n",
       "...        ...                     ...   \n",
       "42861  unknown                 unknown   \n",
       "42867  unknown                 unknown   \n",
       "42869  unknown                 unknown   \n",
       "42872  unknown                 unknown   \n",
       "42876  unknown                 unknown   \n",
       "\n",
       "                                                taxon_id   source          ec  \n",
       "0                                              1886637.0  uniprot     2.6.1.1  \n",
       "1                                              1886637.0  uniprot    4.1.1.12  \n",
       "2                                              1886637.0  uniprot  1.13.11.74  \n",
       "3                                              1886637.0  uniprot  1.13.11.76  \n",
       "4                                              1886637.0  uniprot  1.14.13.23  \n",
       "...                                                  ...      ...         ...  \n",
       "42861  KBase_derived_Bin.028.fastaBA_F_extracted_bins...    KBase    2.7.1.36  \n",
       "42867  KBase_derived_Bin.028.fastaBA_F_extracted_bins...    KBase     6.3.4.3  \n",
       "42869  KBase_derived_Bin.028.fastaBA_F_extracted_bins...    KBase    1.5.98.1  \n",
       "42872  KBase_derived_Bin.028.fastaBA_F_extracted_bins...    KBase     2.7.4.3  \n",
       "42876  KBase_derived_Bin.028.fastaBA_F_extracted_bins...    KBase    2.4.2.48  \n",
       "\n",
       "[1670291 rows x 5 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# dataframes\n",
    "train = pd.read_csv(os.path.join(DATA_DIR, \"taxa2ec-final.csv\"), low_memory=False)\n",
    "test = pd.read_csv(os.path.join(DATA_DIR, 'MAGs', 'FENIX21', 'FENIX21-bins.csv'), low_memory=False)\n",
    "\n",
    "# weird na values, just make some filters\n",
    "test[\"ec\"] = test[\"ec\"].astype(str).dropna()\n",
    "NaN = test[\"ec\"].str.contains(\"NaN\")\n",
    "nan = test[\"ec\"].str.contains(\"nan\")\n",
    "test = test[~NaN & ~nan]\n",
    "\n",
    "# start formatting to merge with taxa2ec-final\n",
    "test[\"media_id\"] = \"unknown\"\n",
    "test[\"species\"] = \"unknown\" ### can update with taxonomy predictions from the KBase workspace\n",
    "test = test.rename(columns={\"filename\": \"taxon_id\", \"ID\": \"node\"})\n",
    "\n",
    "# subset\n",
    "test = test[[\"media_id\",\"species\",\"taxon_id\", \"source\", \"ec\"]]\n",
    "\n",
    "# Concat\n",
    "frames = train, test\n",
    "ml_input = pd.concat(frames)\n",
    "\n",
    "#ml_input.to_csv(os.path.join(DATA_DIR, \"ml_input.csv\"), index=False)\n",
    "ml_input"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "media",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
